assets/js/googlesitekit/data/create-settings-store.js

```diff
@@ -0,0 +1,392 @@
+const RECEIVE_SETTINGS = 'RECEIVE_SETTINGS';
+const RECEIVE_SETTINGS_FAILED = 'RECEIVE_SETTINGS_FAILED';
+const FETCH_SAVE_SETTINGS = 'FETCH_SAVE_SETTINGS';
+const RECEIVE_SAVE_SETTINGS = 'RECEIVE_SAVE_SETTINGS';
+const RECEIVE_SAVE_SETTINGS_FAILED = 'RECEIVE_SAVE_SETTINGS_FAILED';
+
+/**
+ * Creates a store object that includes actions and selectors for managing settings.
+ *
+ * The first three parameters hook up the store to the respective REST API endpoint,
+ * while the fourth defines the names of the sub-settings to support.
+ *
+ * @since n.e.x.t
+ * @private
+ * @param {string} type        The data to access. One of 'core' or 'modules'.
+ * @param {string} identifier  The data identifier, eg. a module slug like 'search-console'.
+ * @param {string} datapoint   The endpoint to request data from, e.g. 'settings'.
+ * @param {Array}  subSettings List of the slugs that are part of the settings object handled
+ *                             by the respective API endpoint.
+ * @return {Object} The settings store object.
+ */
+export const createSettingsStore = ( type, identifier, datapoint, subSettings ) => {
+	const INITIAL_STATE = {
+		settings: undefined,
+		savedSettings: undefined,
+		isFetchingSettings: false,
+		isFetchingSaveSettings: false,
+	};
+
+	// This will be populated further down with sub-setting-specific reducer functions.
+	const subSettingReducers = {};
+
+	const actions = {
+		/**
+		 * Sets settings for the given values.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param {Object} values Settings with their values to set.
+		 * @return {Object} Redux-style action.
+		 */
+		setSettings( values ) {
+			invariant( values, 'values is required.' );
+
+			return {
+				payload: { values },
+				type: SET_SETTINGS,
+			};
+		},
+
+		/**
+		 * Dispatches an action that creates an HTTP request to the settings endpoint.
+		 *
+		 * @since n.e.x.t
+		 * @private
+		 *
+		 * @return {Object} Redux-style action.
+		 */
+		fetchSettings() {
+			return {
+				payload: {},
+				type: FETCH_SETTINGS,
+			};
+		},
+
+		/**
+		 * Stores settings received from the REST API.
+		 *
+		 * @since n.e.x.t
+		 * @private
+		 *
+		 * @param {Array} values Settings with their values from the API.
+		 * @return {Object} Redux-style action.
+		 */
+		receiveSettings( values ) {
+			invariant( values, 'values is required.' );
+
+			return {
+				payload: { values },
+				type: RECEIVE_SETTINGS,
+			};
+		},
+
+		/**
+		 * Dispatches an action signifying the `fetchSettings` side-effect failed.
+		 *
+		 * @since n.e.x.t
+		 * @private
+		 *
+		 * @return {Object} Redux-style action.
+		 */
+		receiveSettingsFailed() {
+			return {
+				payload: {},
+				type: RECEIVE_SETTINGS_FAILED,
+			};
+		},
+
+		/**
+		 * Saves all current settings to the server.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @return {Object} Redux-style action.
+		 */
+		*saveSettings() {
+			try {
+				const savedValues = yield actions.fetchSaveSettings();
```

Comment:
```suggestion
To get store values inside the action you'll need to access the registry, but this is okay 🙂:

```js
				// TODO: We'll need access to the current store name in this code, which we don't right now.
				const updatedSettings = Data.select( STORE_NAME ).getSettings();
				const savedValues = yield actions.fetchSaveSettings( updatedSettings );
```